clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java

```diff
@@ -73,6 +88,74 @@ public void setUp() {
         }
     }
 
+    @Test
+    public void testMemberDataFromSubscription() {
+        List<String> topics = topics(topic);
+        List<TopicPartition> ownedPartitions = partitions(tp(topic1, 0), tp(topic2, 1));
+        List<Subscription> subscriptions = new ArrayList<>();
+        // add all subscription in ConsumerProtocolSubscription versions
+        subscriptions.add(buildSubscriptionV0(topics, ownedPartitions, generationId));
+        subscriptions.add(buildSubscriptionV1(topics, ownedPartitions, generationId));
+        subscriptions.add(buildSubscriptionV2Above(topics, ownedPartitions, generationId));
+        for (Subscription subscription : subscriptions) {
+            if (subscription != null) {
+                AbstractStickyAssignor.MemberData memberData = assignor.memberDataFromSubscription(subscription);
+                assertEquals(ownedPartitions, memberData.partitions, "subscription: " + subscription + " doesn't have expected owned partition");
+                assertEquals(generationId, memberData.generation.orElse(-1), "subscription: " + subscription + " doesn't have expected generation id");
+            }
+        }
+    }
+
+    @Test
+    public void testMemberDataFromSubscriptionWithInconsistentData() {
+        Map<String, Integer> partitionsPerTopic = new HashMap<>();
+        partitionsPerTopic.put(topic, 2);
+        List<TopicPartition> ownedPartitionsInUserdata = partitions(tp1);
+        List<TopicPartition> ownedPartitionsInSubscription = partitions(tp0);
+
+        assignor.onAssignment(new ConsumerPartitionAssignor.Assignment(ownedPartitionsInUserdata), new ConsumerGroupMetadata(groupId, generationId, consumer1, Optional.empty()));
+        ByteBuffer userDataWithHigherGenerationId = assignor.subscriptionUserData(new HashSet<>(topics(topic)));
+        // The owned partitions and generation id are provided in user data and different owned partition is provided in subscription without generation id
+        // If subscription provides no generation id, we'll honor the generation id in userData and owned partitions in subscription
+        Subscription subscription = new Subscription(topics(topic), userDataWithHigherGenerationId, ownedPartitionsInSubscription);
+
+        AbstractStickyAssignor.MemberData memberData = assignor.memberDataFromSubscription(subscription);
+        // in CooperativeStickyAssignor, we'll serialize owned partition in subscription into userData
+        // but in StickyAssignor, we'll serialize owned partition in assignment into userData
+        if (assignor instanceof CooperativeStickyAssignor) {
```

Comment:
This is not very elegant given that we have subclasses for each. Should we move this test to the respective subclasses?